framework/src/problems/SubProblem.C

```diff
@@ -967,15 +967,13 @@ SubProblem::addAlgebraicGhostingFunctor(GhostingFunctor & algebraic_gf, bool to_
 }
 
 void
-SubProblem::addAlgebraicGhostingFunctor(std::shared_ptr<GhostingFunctor> algebraic_gf, bool to_mesh)
+SubProblem::addCouplingFunctor(GhostingFunctor & coupling_functor, bool to_mesh)
 {
-  EquationSystems & eq = es();
-  const auto n_sys = eq.n_systems();
-  if (!n_sys)
-    return;
-
-  eq.get_system(0).get_dof_map().add_algebraic_ghosting_functor(algebraic_gf, to_mesh);
-  cloneAlgebraicGhostingFunctor(*algebraic_gf, to_mesh);
+  for (const auto nl_sys_num : make_range(numNonlinearSystems()))
+  {
+    auto & nl = systemBaseNonlinear(nl_sys_num);
+    nl.system().get_dof_map().add_coupling_functor(coupling_functor, to_mesh);
+  }
```

Comment:
```suggestion
  const auto num_nl_sys = numNonlinearSystems();
  if (!num_nl_sys)
    return;

  systemBaseNonlinear(0).system().get_dof_map().add_coupling_functor(coupling_functor, to_mesh);

  if (num_nl_sys == 1)
    return;

  auto pr = _root_coupling_functor_to_sys_clones.emplace(
      &coupling_functor, std::vector<std::shared_ptr<GhostingFunctor>>(num_nl_sys - 1));
  mooseAssert(pr.second, "We are adding a duplicate coupling functor");
  auto & clones_vec = pr.first->second;

  for (const auto i : make_range(1, num_nl_sys))
  {
    DofMap & dof_map = systemBaseNonlinear(i).system().get_dof_map();
    std::shared_ptr<GhostingFunctor> clone_coupling_functor = coupling_functor.clone();
    std::dynamic_pointer_cast<RelationshipManager>(clone_coupling_functor)
        ->init(*coupling_functor.get_mesh(), &dof_map);
    dof_map.add_coupling_functor(clone_coupling_functor, to_mesh);
    clones_vec[i - 1] = clone_coupling_functor;
  }
```
I didn't think critically enough at first. We've had problems in the past when you add a ghosting functor to a DofMap (or mesh) where the ghosting functor actually has a different DofMap (or mesh) associated with it. We have to make sure we are consistent with the ghosting-functor/DofMap (or mesh) relationship.

This change will necessitate adding a `_root_coupling_functor_to_sys_clones` data member in `SubProblem` (like we have for the algebraic ghosting functors)

Reply:
Got it thanks Alex!